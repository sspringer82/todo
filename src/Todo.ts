export type Todo = {
    id: number,
    title: string,
    done: boolean,
    category?: string,
    color?: string,
    comment?: string,
    subtask?: Subtask[]
};

export type TodoInput = {id?: number} & Omit<Todo, 'id'>;

export type Subtask = {todoId: number} & Omit<Todo, 'comment' | 'subtask'>

export type SubtaskInput = {id?: number} & Omit<Subtask, 'id'>

export const initialTodo:TodoInput = {
    title: '',
    done: false,
    comment: ''
};
Object.freeze(initialTodo);