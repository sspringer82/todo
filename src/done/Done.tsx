import React from "react";
import { Todo, TodoInput } from "../Todo";
import produce from "immer";
import ClearIcon from "@material-ui/icons/Clear";
import CheckIcon from "@material-ui/icons/Check";

export type Props = {
  todo: Todo;
  onSave(todo: TodoInput): Promise<void>;
};

const Done: React.FC<Props> = ({ todo, onSave }) => {
  function handleStatusToggle(todo: Todo) {
    onSave(
      produce(todo, (draftTodo) => {
        draftTodo.done = !draftTodo.done;
      })
    );
  }

  return (
    <div>
      <button
        className="focus:outline-none"
        onClick={() => handleStatusToggle(todo)}
        data-testid="done-button"
      >
        {todo.done ? (
          <CheckIcon style={{ color: "limegreen" }} data-testid="done-done" />
        ) : (
          <ClearIcon style={{ color: "red" }} data-testid="done-notdone" />
        )}
      </button>
    </div>
  );
};

export default Done;
