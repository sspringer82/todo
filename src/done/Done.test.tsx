import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Done, { Props } from './Done';
import { Todo } from '../Todo';

describe('Done', () => {

  let props: Props;

  beforeEach(() => {
    props = {
      onSave: jest.fn(),
      todo:  {
        id: 1,
        title: 'Test',
        done: false
      }
    }
  });

  it("should perform a snapshot test", () => {
    const {container} = render(<Done {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should toggle a not done task to done', () => {
    const {getByTestId} = render(<Done {...props} />);
    fireEvent.click(getByTestId('done-button'));

    const expected = {
      id: 1,
      title: 'Test',
      done: true
    }
    expect(props.onSave).toHaveBeenCalledWith(expected);
  });

  it('should toggle a done task to not done', () => {
    props.todo.done = true;
    
    const {getByTestId} = render(<Done {...props} />);
    fireEvent.click(getByTestId('done-button'));

    const expected = {
      id: 1,
      title: 'Test',
      done: false
    }
    expect(props.onSave).toHaveBeenCalledWith(expected);
  });

  it('should have the done icon if done', () => {
    props.todo.done = true;
    
    const {getByTestId, queryByTestId} = render(<Done {...props} />);

    const done = getByTestId('done-done');
    const notDone = queryByTestId('done-notdone');

    expect(done).toBeInTheDocument();
    expect(notDone).toBeNull();
  });

  it('should have the notdone icon if notdone', () => {
    const {getByTestId, queryByTestId} = render(<Done {...props} />);

    const notDone = getByTestId('done-notdone');
    const done = queryByTestId('done-done');

    expect(notDone).toBeInTheDocument();
    expect(done).toBeNull();
  });
});
