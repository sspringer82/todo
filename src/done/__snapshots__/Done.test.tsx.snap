// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Done should perform a snapshot test 1`] = `
<div>
  <div>
    <button
      class="focus:outline-none relative"
      data-testid="done-button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        data-testid="done-notdone"
        focusable="false"
        style="color: red;"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;
